src/callbacks_step/analysis.jl

```diff
@@ -184,175 +188,208 @@ function (analysis_callback::AnalysisCallback)(integrator)
+      # L2 error
+      if :l2_error_primitive in analysis_errors
+        print(" L2 error prim.: ")
         for v in eachvariable(equations)
-          mpi_isroot() && @printf("   %-14s", varnames(cons2prim, equations)[v])
-        end
-        mpi_println()
-
-        # L2 error
-        if :l2_error_primitive in analysis_errors
-          mpi_print(" L2 error prim.: ")
-          for v in eachvariable(equations)
-            mpi_isroot() && @printf("%10.8e   ", l2_error_prim[v])
-            analysis_callback.save_analysis && mpi_isroot() && @printf(io, "  % 10.8e", l2_error_prim[v])
-          end
-          mpi_println()
+          @printf("%10.8e   ", l2_error_prim[v])
+          @printf(io, "  % 10.8e", l2_error_prim[v])
         end
+        println()
+      end
 
-        # L∞ error
-        if :linf_error_primitive in analysis_errors
-          mpi_print(" Linf error pri.:")
-          for v in eachvariable(equations)
-            mpi_isroot() && @printf("%10.8e   ", linf_error_prim[v])
-            analysis_callback.save_analysis && mpi_isroot() && @printf(io, "  % 10.8e", linf_error_prim[v])
-          end
-          mpi_println()
+      # L∞ error
+      if :linf_error_primitive in analysis_errors
+        print(" Linf error pri.:")
+        for v in eachvariable(equations)
+          @printf("%10.8e   ", linf_error_prim[v])
+          @printf(io, "  % 10.8e", linf_error_prim[v])
         end
+        println()
       end
+    end
+  end
 
+  # additional integrals
+  analyze_integrals(analysis_integrals, io, du, u, t, semi)
 
-      # additional
-      for quantity in analysis_integrals
-        res = analyze(quantity, du, u, t, semi)
-        mpi_isroot() && @printf(" %-12s:", pretty_form_utf(quantity))
-        mpi_isroot() && @printf("  % 10.8e", res)
-        analysis_callback.save_analysis && mpi_isroot() && @printf(io, "  % 10.8e", res)
-        mpi_println()
-      end
-    end # GC.@preserve du_ode
+  return l2_error, linf_error
+end
 
-    mpi_println("─"^100)
-    mpi_println()
 
-    # Add line break and close analysis file if it was opened
-    if mpi_isroot() && analysis_callback.save_analysis
-      println(io)
-      close(io)
+# Print level information only if AMR is enabled
+function print_amr_information(callbacks, mesh, solver, cache)
+
+  if uses_amr(callbacks)
+    levels = Vector{Int}(undef, nelements(solver, cache))
+    min_level = typemax(Int)
+    max_level = typemin(Int)
+    for element in eachelement(solver, cache)
+      current_level = mesh.tree.levels[cache.elements.cell_ids[element]]
+      levels[element] = current_level
+      min_level = min(min_level, current_level)
+      max_level = max(max_level, current_level)
+    end
+
+    for level = max_level:-1:min_level+1
+      mpi_println(" ├── level $level:    " * @sprintf("% 14d", count(isequal(level), levels)))
     end
+    mpi_println(" └── level $min_level:    " * @sprintf("% 14d", count(isequal(min_level), levels)))
   end
```

Comment:
Maybe like this to avoid indenting the entire function body?
```suggestion
  # Return early if there is nothing to print
  if !uses_amr(callbacks)
    return nothing
  end
  
  levels = Vector{Int}(undef, nelements(solver, cache))
  min_level = typemax(Int)
  max_level = typemin(Int)
  for element in eachelement(solver, cache)
    current_level = mesh.tree.levels[cache.elements.cell_ids[element]]
    levels[element] = current_level
    min_level = min(min_level, current_level)
    max_level = max(max_level, current_level)
  end

  for level = max_level:-1:min_level+1
    mpi_println(" ├── level $level:    " * @sprintf("% 14d", count(isequal(level), levels)))
  end
  mpi_println(" └── level $min_level:    " * @sprintf("% 14d", count(isequal(min_level), levels)))
end
```

Reply:
Thanks! Done in 21167306